pkg/k8s/cnpstatus.go

```diff
@@ -0,0 +1,301 @@
+	cnpKey := path.Join(string(cnp.UID), cnp.Namespace, cnp.Name)
+	ch, ok := c.eventMap.createIfNotExist(cnpKey)
+	if ok {
+		return
+	}
+
+	namespace := cnp.Namespace
+	name := cnp.Name
+
+	nodeStatusMap := make(map[string]cilium_v2.CiliumNetworkPolicyNodeStatus)
+	var hasRun bool
+	c.controllers.UpdateController(cnpKey, controller.ControllerParams{
+		DoFunc: func(ctx context.Context) error {
+			// Get any updates received in case other nodes updated the kvstore
+			// before this controller was started, as cilium-agents running in
+			// the cluster may receive the CNP event from their own Kubernetes
+			// watchers and send keys for their status to the key-value store
+			// before this controller has been started.
+			if !hasRun {
+				sharedKeys := c.cnpStore.SharedKeysMap()
+				for k := range sharedKeys {
+					// Look for any key which matches this CNP.
+					if strings.HasPrefix(k, cnpKey) {
+						v, ok := sharedKeys[k].(*CNPNSWithMeta)
+						if !ok {
+							log.Errorf("received unexpected type mapping to key %s in cnp shared store: %T", k, sharedKeys[k])
+							continue
+						}
+						nodeStatusMap[v.Node] = v.CiliumNetworkPolicyNodeStatus
+					}
+				}
+				hasRun = true
+			} else {
+				// Wait for one status update to occur.
+				select {
+				case ev, ok := <-ch:
+					if ok {
+						nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
+					}
+				case <-ctx.Done():
+					return nil
+				}
+			}
+
+		Loop:
+			for {
+				select {
+				case <-ctx.Done():
+					// Controller was stopped, we can simply exit.
+					return nil
+				case ev, ok := <-ch:
+					if ok {
+						nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
+					}
+				default:
+					// Nothing to do, update K8s apiserver with status updates.
+					break Loop
+				}
+			}
+
+			// This will only be the case after the first run has occurred and
+			// there are no subsequent status updates since we don't want for
+			// at least one event in the first run.
+			if len(nodeStatusMap) == 0 {
+				return nil
+			}
+
+			var (
+				cnp *types.SlimCNP
+				err error
+			)
+
+			switch {
+			// Patching doesn't need us to get the CNP from
+			// the store because we can perform patches without
+			// needing the actual CNP object itself.s
+			case k8sversion.Capabilities().Patch:
+			default:
+				cnp, err = getUpdatedCNPFromStore(c.k8sStore, fmt.Sprintf("%s/%s", namespace, name))
+				if err != nil {
+					log.WithError(err).Error("error getting updated cnp from store")
+					return err
+				}
+			}
+
+			// Now that we have collected all events for
+			// the given CNP, update the status for all nodes
+			// which have sent us updates.
+			if err = updateStatusesByCapabilities(CiliumClient(), k8sversion.Capabilities(), cnp, namespace, name, nodeStatusMap); err != nil {
+				return err
+			}
+			return nil
+		},
+		RunInterval: c.updateInterval,
+		StopFunc: func(ctx context.Context) error {
+			close(ch)
```

Comment:
Closing this here will make panics in [1]

Reply:
I've added a channel which is closed upon receiving a delete event for a given CNP. The kvstore watcher, which is the sender to the status update channel, now also closes the channel if this shutdown channel is closed, ensuring no panics will occur. The goroutine which is responsible for receiving status update events also will exit if this channel is closed.